kv/kvserver: TestDefaultConnectionDisruptionDoesNotInterfereWithSystemTraffic failed

kv/kvserver.TestDefaultConnectionDisruptionDoesNotInterfereWithSystemTraffic [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/8228444?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/8228444?buildTab=artifacts#/) on master @ [7594fb2d1438dd5ed8f80a887b8d60b6cdd296ed](https://github.com/cockroachdb/cockroach/commits/7594fb2d1438dd5ed8f80a887b8d60b6cdd296ed):


```
=== RUN   TestDefaultConnectionDisruptionDoesNotInterfereWithSystemTraffic
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/9ded2e1b6e3cd0145900033ce06c11b0/logTestDefaultConnectionDisruptionDoesNotInterfereWithSystemTraffic3648243196
    test_log_scope.go:79: use -show-logs to present logs inline
=== CONT  TestDefaultConnectionDisruptionDoesNotInterfereWithSystemTraffic
    client_raft_test.go:4950: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/9ded2e1b6e3cd0145900033ce06c11b0/logTestDefaultConnectionDisruptionDoesNotInterfereWithSystemTraffic3648243196
--- FAIL: TestDefaultConnectionDisruptionDoesNotInterfereWithSystemTraffic (11.21s)
=== RUN   TestDefaultConnectionDisruptionDoesNotInterfereWithSystemTraffic/after_restart
    client_raft_test.go:4927: 
        	Error Trace:	/home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/1413/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/kv/kvserver/kvserver_test_/kvserver_test.runfiles/com_github_cockroachdb_cockroach/pkg/kv/kvserver/client_raft_test.go:4927
        	Error:      	Received unexpected error:
        	            	aborted during DistSender.Send: context deadline exceeded
        	            	(1)
        	            	  | (opaque error wrapper)
        	            	  | type name: github.com/cockroachdb/errors/withstack/*withstack.withStack
        	            	  | reportable 0:
        	            	  |
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*DistSender).sendToReplicas
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/dist_sender.go:2375
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*DistSender).sendPartialBatch
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/dist_sender.go:1668
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*DistSender).divideAndSendBatchToRanges
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/dist_sender.go:1240
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*DistSender).Send
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/dist_sender.go:861
        	            	  | github.com/cockroachdb/cockroach/pkg/kv.(*CrossRangeTxnWrapperSender).Send
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/db.go:223
        	            	  | github.com/cockroachdb/cockroach/pkg/kv.(*DB).sendUsingSender
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/db.go:998
        	            	  | github.com/cockroachdb/cockroach/pkg/kv.(*DB).send
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/db.go:981
        	            	  | github.com/cockroachdb/cockroach/pkg/kv.sendAndFill
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/db.go:835
        	            	  | github.com/cockroachdb/cockroach/pkg/kv.(*DB).Run
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/db.go:858
        	            	  | github.com/cockroachdb/cockroach/pkg/kv.(*DB).Put
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/db.go:380
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.TestDefaultConnectionDisruptionDoesNotInterfereWithSystemTraffic.func2
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/client_raft_test.go:4927
        	            	  | testing.tRunner
        	            	  | 	GOROOT/src/testing/testing.go:1446
        	            	  | runtime.goexit
        	            	  | 	GOROOT/src/runtime/asm_arm64.s:1172
        	            	Wraps: (2) aborted during DistSender.Send
        	            	Wraps: (3) context deadline exceeded
        	            	Error types: (1) *errbase.opaqueWrapper (2) *errutil.withPrefix (3) context.deadlineExceededError
        	Test:       	TestDefaultConnectionDisruptionDoesNotInterfereWithSystemTraffic/after_restart
        	Messages:   	Expected success writing to liveness range
    --- FAIL: TestDefaultConnectionDisruptionDoesNotInterfereWithSystemTraffic/after_restart (2.09s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/replication
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestDefaultConnectionDisruptionDoesNotInterfereWithSystemTraffic.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-23197

kv/kvserver.TestDefaultConnectionDisruptionDoesNotInterfereWithSystemTraffic [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/8998854?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/8998854?buildTab=artifacts#/) on master @ [01e14f68f7c757cab1c878cebfecd73c5119046b](https://github.com/cockroachdb/cockroach/commits/01e14f68f7c757cab1c878cebfecd73c5119046b):


```
=== RUN   TestDefaultConnectionDisruptionDoesNotInterfereWithSystemTraffic
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/33e1d369c27b9c01b2b6009c561815a3/logTestDefaultConnectionDisruptionDoesNotInterfereWithSystemTraffic2492123383
    test_log_scope.go:79: use -show-logs to present logs inline
=== CONT  TestDefaultConnectionDisruptionDoesNotInterfereWithSystemTraffic
    client_raft_test.go:4949: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/33e1d369c27b9c01b2b6009c561815a3/logTestDefaultConnectionDisruptionDoesNotInterfereWithSystemTraffic2492123383
--- FAIL: TestDefaultConnectionDisruptionDoesNotInterfereWithSystemTraffic (17.14s)
=== RUN   TestDefaultConnectionDisruptionDoesNotInterfereWithSystemTraffic/after_restart
    client_raft_test.go:4926: 
        	Error Trace:	/home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/1220/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/kv/kvserver/kvserver_test_/kvserver_test.runfiles/com_github_cockroachdb_cockroach/pkg/kv/kvserver/client_raft_test.go:4926
        	Error:      	Received unexpected error:
        	            	aborted during DistSender.Send: context deadline exceeded
        	            	(1)
        	            	  | (opaque error wrapper)
        	            	  | type name: github.com/cockroachdb/errors/withstack/*withstack.withStack
        	            	  | reportable 0:
        	            	  |
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*DistSender).sendToReplicas
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/dist_sender.go:2390
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*DistSender).sendPartialBatch
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/dist_sender.go:1668
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*DistSender).divideAndSendBatchToRanges
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/dist_sender.go:1239
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*DistSender).Send
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/dist_sender.go:860
        	            	  | github.com/cockroachdb/cockroach/pkg/kv.(*CrossRangeTxnWrapperSender).Send
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/db.go:223
        	            	  | github.com/cockroachdb/cockroach/pkg/kv.(*DB).sendUsingSender
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/db.go:1001
        	            	  | github.com/cockroachdb/cockroach/pkg/kv.(*DB).send
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/db.go:984
        	            	  | github.com/cockroachdb/cockroach/pkg/kv.sendAndFill
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/db.go:831
        	            	  | github.com/cockroachdb/cockroach/pkg/kv.(*DB).Run
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/db.go:854
        	            	  | github.com/cockroachdb/cockroach/pkg/kv.(*DB).Put
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/db.go:380
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.TestDefaultConnectionDisruptionDoesNotInterfereWithSystemTraffic.func2
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/client_raft_test.go:4926
        	            	  | testing.tRunner
        	            	  | 	GOROOT/src/testing/testing.go:1446
        	            	  | runtime.goexit
        	            	  | 	src/runtime/asm_amd64.s:1594
        	            	Wraps: (2) aborted during DistSender.Send
        	            	Wraps: (3) context deadline exceeded
        	            	Error types: (1) *errbase.opaqueWrapper (2) *errutil.withPrefix (3) context.deadlineExceededError
        	Test:       	TestDefaultConnectionDisruptionDoesNotInterfereWithSystemTraffic/after_restart
        	Messages:   	Expected success writing to liveness range
    --- FAIL: TestDefaultConnectionDisruptionDoesNotInterfereWithSystemTraffic/after_restart (1.02s)
```
<p>Parameters: <code>TAGS=bazel,gss,race</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestDefaultConnectionDisruptionDoesNotInterfereWithSystemTraffic.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/kv/kvserver/client_raft_test.go
+++ b/pkg/kv/kvserver/client_raft_test.go
@@ -4834,7 +4834,6 @@ func (cs *disablingClientStream) SendMsg(m interface{}) error {
 // traffic on the SystemClass connection.
 func TestDefaultConnectionDisruptionDoesNotInterfereWithSystemTraffic(t *testing.T) {
 	defer leaktest.AfterTest(t)()
-	skip.WithIssue(t, 94875, "flaky test")
 	defer log.Scope(t).Close(t)
 
 	stickyEngineRegistry := server.NewStickyInMemEnginesRegistry()
@@ -4918,30 +4917,47 @@ func TestDefaultConnectionDisruptionDoesNotInterfereWithSystemTraffic(t *testing
 		tc.WaitForValues(t, keyA, []int64{1, 1, 1})
 		disabled.Store(true)
 
-		// Set a relatively short timeout so that this test doesn't take too long.
-		// We should always hit it.
-		withTimeout, cancel := context.WithTimeout(ctx, 1*time.Second)
-		defer cancel()
+		// Create a pair of utilities to perform operations under suitable timeouts.
+		expSucceed := func(msg string, fn func(context.Context) error) {
+			// Set a long timeout so that this test doesn't flake under stress.
+			// We should never hit it.
+			ctxTimeout, cancel := context.WithTimeout(ctx, testutils.DefaultSucceedsSoonDuration)
+			defer cancel()
+			require.NoError(t, fn(ctxTimeout), "Expected success "+msg)
+		}
+		expTimeout := func(msg string, fn func(context.Context) error) {
+			// Set a relatively short timeout so that this test doesn't take too long.
+			// We should always hit it.
+			ctxTimeout, cancel := context.WithTimeout(ctx, 100*time.Millisecond)
+			defer cancel()
+			require.ErrorIs(t, fn(ctxTimeout), context.DeadlineExceeded, "Expected timeout "+msg)
+		}
 
 		// Write to the liveness range on the System class.
-		require.NoError(t, db.Put(withTimeout, keyLiveness, 2), "Expected success writing to liveness range")
+		expSucceed("writing to liveness range", func(ctx context.Context) error {
+			return db.Put(ctx, keyLiveness, 2)
+		})
 
 		// Write to the standard range on the default class.
-		require.ErrorIs(t, db.Put(withTimeout, keyA, 2), context.DeadlineExceeded,
-			"Expected timeout writing to key range")
+		expTimeout("writing to key range", func(ctx context.Context) error {
+			return db.Put(ctx, keyA, 2)
+		})
 
 		// Write to the liveness range on the System class with system disabled to
 		// ensure the test is actually working.
 		disabledSystem.Store(true)
-		require.ErrorIs(t, db.Put(withTimeout, keyLiveness, 2),
-			context.DeadlineExceeded, "Expected timeout writing to liveness range")
+		expTimeout("writing to liveness range", func(ctx context.Context) error {
+			return db.Put(ctx, keyLiveness, 2)
+		})
 		disabledSystem.Store(false)
 
-		// Heal the partition, the previous proposal may now succeed but it may have
+		// Heal the partition, the previous proposal may now succeed but it also may
 		// have been canceled.
 		disabled.Store(false)
 		// Overwrite with a new value and ensure that it propagates.
-		require.NoError(t, db.Put(ctx, keyA, 3), "Expected success after healed partition")
+		expSucceed("after healed partition", func(ctx context.Context) error {
+			return db.Put(ctx, keyA, 3)
+		})
 		tc.WaitForValues(t, keyA, []int64{3, 3, 3})
 	}
 	t.Run("initial_run", runTest)

